src/com/facebook/buck/android/MergeAndroidResourcesStep.java

```diff
@@ -346,26 +351,77 @@ void writePerPackageRDotJava(
       }
 
       String packageName = entry.getValue();
-      for (String line : linesInSymbolsFile) {
-        Optional<RDotTxtEntry> parsedEntry = RDotTxtEntry.parse(line);
-        Preconditions.checkState(parsedEntry.isPresent(), "Should be able to match '%s'.", line);
-
-        // We're only doing the remapping so Roboelectric is happy and it is already ignoring the
-        // id references found in the styleable section.  So let's do that as well so we don't have
-        // to get fancier than is needed.  That is, just re-enumerate all app-level resource ids
-        // and ignore everything else, allowing the styleable references to be messed up.
-        RDotTxtEntry resource = parsedEntry.get();
+
+      int index = 0;
+      while (index < linesInSymbolsFile.size()) {
+        RDotTxtEntry resource = getResourceAtIndex(linesInSymbolsFile, index);
+        index++;
+
         if (uberRDotTxtIds.isPresent()) {
           Preconditions.checkNotNull(finalIds);
           if (!finalIds.containsKey(resource)) {
             LOG.debug("Cannot find resource '%s' in the uber R.txt.", resource);
             continue;
           }
           resource = resource.copyWithNewIdValue(finalIds.get(resource));
-        } else if (resource.idValue.startsWith("0x7f")) {
-          Preconditions.checkNotNull(enumerator);
-          resource = resource.copyWithNewIdValue(String.format("0x%08x", enumerator.next()));
+        } else {
+          if (resourceToIdValuesMap.get(resource) != null) {
+            resource = resource.copyWithNewIdValue(resourceToIdValuesMap.get(resource));
+          } else {
+
+            if (resource.idType == IdType.INT_ARRAY &&
+                resource.type == RType.STYLEABLE) {
+
+              List<String> styleableIdValues = new ArrayList<>();
+              int styleableIndex = 0;
+
+              while (styleableIndex + index < linesInSymbolsFile.size()) {
```

Comment:
this is still O(N). the outer loop goes from 0 to linesInSymbolsFile.size() & the inner loop goes through styleable int entries once more which is at-most N.

will move this to a multimap.